counter/static/webpack/graph-index.ts

```diff
@@ -0,0 +1,152 @@
+    if (options.startDate <= info.start) {
+      return;
+    }
+    const newPerms = await paginated(permanencyFetchPermanancies, {
+      query: {
+        counter: [options.counterId],
+        // biome-ignore lint/style/useNamingConvention: backend API uses snake_case
+        end_after: info.startStr,
+        // biome-ignore lint/style/useNamingConvention: backend API uses snake_case
+        start_before: info.endStr,
+      },
+    } as PermanencyFetchPermananciesData);
+    options.startDate = info.start;
+    calendar.addEventSource(getEvents(newPerms, false));
+    permanancies.push(...newPerms);
+    calendar.render();
+  });
+}
+
+function roundToQuarter(date: Date, ceil: boolean) {
+  const result = date;
+  const minutes = date.getMinutes();
+  // removes minutes exceeding the lower quarter and adds 15 minutes if rounded to ceiling
+  result.setMinutes(minutes - (minutes % 15) + +ceil * 15, 0, 0);
+  return result;
+}
+
+function convertPermanancyToOpeningTime(permanancy: PermanencySchema): OpeningTime {
+  return {
+    start: roundToQuarter(new Date(permanancy.start), false),
+    end: roundToQuarter(new Date(permanancy.end ?? Date.now()), true),
+  };
+}
+
+function getOpeningTimes(rawPermanancies: PermanencySchema[]) {
+  const permanancies = rawPermanancies
+    .map(convertPermanancyToOpeningTime)
+    .sort((a, b) => a.start.getTime() - b.start.getTime());
+
+  const openingTimes: OpeningTime[] = [];
+
+  for (const permanancy of permanancies) {
+    // if there are no opening times, add the first one
+    if (openingTimes.length === 0) {
+      openingTimes.push(permanancy);
+    } else {
+      const lastPermanancy = openingTimes[openingTimes.length - 1];
+      // if the new permanancy starts before the 15 minutes following the end of the last one, merge them
+      if (permanancy.start <= lastPermanancy.end) {
+        lastPermanancy.end = new Date(
+          Math.max(lastPermanancy.end.getTime(), permanancy.end.getTime()),
+        );
+      } else {
+        openingTimes.push(permanancy);
+      }
+    }
+  }
+  return openingTimes;
+}
+
+function getEvents(permanancies: PermanencySchema[], currentWeek = true): EventInput[] {
+  const openingTimes = getOpeningTimes(permanancies);
+  const events: EventInput[] = [];
+  for (const openingTime of openingTimes) {
+    let shift = false;
+    if (currentWeek) {
+      const lastMonday = getLastMonday();
+      shift = openingTime.end < lastMonday;
+    }
+    // if permanancies took place last week (=before monday),
+    // -> display them in lightblue as part of the current week
+    events.push({
+      start: shift ? shiftDateByDays(openingTime.start, 7) : openingTime.start,
+      end: shift ? shiftDateByDays(openingTime.end, 7) : openingTime.end,
+      backgroundColor: shift ? "lightblue" : "green",
+    });
+  }
+  return events;
+}
+
+// Function to get last Monday at 00:00
+function getLastMonday(now = new Date()): Date {
```

Comment:
Ca pose pas de problème en JS, de mettre un appel de fonction/constructeur en paramètre ?

Reply:
non pas du tout, ça fait juste que le paramètre est optionnel, et prends une valeur par défaut a un nouvel objet, en l'occurrence la date du moment présent
c'est aussi faisable en C++ ce genre de trucs

Reply:
Faut se méfier du "ça se fait aussi dans tel langage", parce que justement en Python ça se fait pas. On a même eu des problèmes dans le passé avec ça (cf #656) et Ruff a un lint pour ce cas très précis (https://docs.astral.sh/ruff/rules/function-call-in-default-argument/).